File: tools/validate.js
// questions.js 검증: 총 문항수, 선택지 5개, answer 0~4, PDF당 10문제 이상, 중복 id
const fs = require("fs");
const path = require("path");
global.window = {};
require(path.join(__dirname, "..", "questions.js"));
const Q = global.window.QUESTIONS;

let errors = [];
const ids = new Set();
const perSource = {};   // week-session -> count
const typeCount = {};

for (const q of Q) {
  if (ids.has(q.id)) errors.push(`중복 id: ${q.id}`);
  ids.add(q.id);
  if (!Array.isArray(q.choices) || q.choices.length !== 5)
    errors.push(`${q.id}: choices 개수 ${q.choices ? q.choices.length : "없음"}`);
  if (typeof q.answer !== "number" || q.answer < 0 || q.answer > 4)
    errors.push(`${q.id}: answer 범위 오류 (${q.answer})`);
  if (!q.question || !q.explanation || !q.source)
    errors.push(`${q.id}: 필수 필드 누락`);
  const key = `w${q.week}-s${q.session}`;
  perSource[key] = (perSource[key] || 0) + 1;
  typeCount[q.type] = (typeCount[q.type] || 0) + 1;
}

console.log(`총 문항 수: ${Q.length}`);
console.log(`고유 id 수: ${ids.size}`);
console.log(`\n유형별 분포:`);
for (const t of Object.keys(typeCount)) {
  console.log(`  ${t}: ${typeCount[t]} (${Math.round(typeCount[t]/Q.length*100)}%)`);
}
console.log(`\nPDF(주차-차시)별 문항 수:`);
const lowSources = [];
for (const k of Object.keys(perSource).sort()) {
  const n = perSource[k];
  console.log(`  ${k}: ${n}`);
  if (k !== "w16-s0" && n < 10) lowSources.push(`${k}(${n})`);
}

console.log(`\n=== 검증 결과 ===`);
if (lowSources.length) errors.push(`10문제 미만 PDF: ${lowSources.join(", ")}`);
if (errors.length) {
  console.log("❌ 오류:");
  errors.forEach(e => console.log("  - " + e));
  process.exit(1);
} else {
  console.log("✅ 모든 검증 통과");
}

File: tools/validate.py
#!/usr/bin/env python3
"""questions.js 검증 (node 없이 파이썬으로). 따옴표를 인식해 선택지 개수를 정확히 센다."""
import re, os, sys

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
src = open(os.path.join(ROOT, "questions.js"), encoding="utf-8").read()

# 각 문항 객체는 '{ id:' 로 시작
blocks = re.split(r'\n\{ id:', src)
blocks = ['{ id:' + b for b in blocks[1:]]  # 첫 조각(헤더) 제외

def count_choices(block):
    m = re.search(r'choices:\[', block)
    if not m:
        return -1
    i = m.end()
    depth = 1
    in_str = False
    esc = False
    items = 0
    started = False
    while i < len(block) and depth > 0:
        c = block[i]
        if in_str:
            if esc: esc = False
            elif c == '\\': esc = True
            elif c == '"': in_str = False
        else:
            if c == '"': in_str = True; started = True
            elif c == '[': depth += 1
            elif c == ']':
                depth -= 1
            elif c == ',' and depth == 1:
                items += 1
        i += 1
    return items + 1 if started else 0

errors = []
ids = set()
per_source = {}
type_count = {}
total = 0

for b in blocks:
    total += 1
    mid = re.search(r'id:"([^"]+)"', b)
    qid = mid.group(1) if mid else f"(unknown@{total})"
    if qid in ids: errors.append(f"중복 id: {qid}")
    ids.add(qid)
    week = int(re.search(r'week:(\d+)', b).group(1))
    session = int(re.search(r'session:(\d+)', b).group(1))
    answer = int(re.search(r'answer:(\d+)', b).group(1))
    nch = count_choices(b)
    if nch != 5: errors.append(f"{qid}: choices 개수 {nch}")
    if not (0 <= answer <= 4): errors.append(f"{qid}: answer 범위 {answer}")
    for fld in ("question:", "explanation:", "source:"):
        if fld not in b: errors.append(f"{qid}: {fld} 누락")
    key = f"w{week:02d}-s{session}"
    per_source[key] = per_source.get(key, 0) + 1
    t = re.search(r'type:"([^"]+)"', b)
    tt = t.group(1) if t else "?"
    type_count[tt] = type_count.get(tt, 0) + 1

print(f"총 문항 수: {total}")
print(f"고유 id 수: {len(ids)}")
print("\n유형별 분포:")
for t in sorted(type_count):
    print(f"  {t}: {type_count[t]} ({round(type_count[t]/total*100)}%)")
print("\nPDF(주차-차시)별 문항 수:")
low = []
for k in sorted(per_source):
    n = per_source[k]
    print(f"  {k}: {n}")
    if k != "w16-s0" and n < 10:
        low.append(f"{k}({n})")

print("\n=== 검증 결과 ===")
if low: errors.append("10문제 미만 PDF: " + ", ".join(low))
if errors:
    print("[X] 오류:")
    for e in errors: print("  - " + e)
    sys.exit(1)
else:
    print("[OK] 모든 검증 통과")
